akka-http2-support/src/test/scala/akka/http/impl/engine/http2/TelemetrySpiSpec.scala

```diff
@@ -0,0 +1,237 @@
+          .toMat(Sink.actorRef(probe.ref, "done"))(Keep.both)
+          .run()
+      requestQueue.offer(HttpRequest())
+
+      probe.expectMsg("seen-connection")
+      probe.expectMsg("request-seen")
+      probe.expectMsgType[TelemetryAttributes.ClientMeta]
+      val requestId = probe.expectMsgType[RequestId]
+      val requestIdOnServer = probe.expectMsgType[String]
+      requestIdOnServer should ===(requestId.id)
+
+      probe.expectMsg("response-seen")
+      val responseId = probe.expectMsgType[RequestId]
+      requestId should ===(responseId)
+      val response = probe.expectMsgType[HttpResponse]
+      response.attribute(requestIdAttr) should be(None)
+      requestQueue.complete()
+
+      probe.expectMsg("close-seen")
+      serverBinding.terminate(3.seconds).futureValue
+    }
+
+    "hook into HTTP2 server requests" in {
+      val probe = TestProbe()
+      TestTelemetryImpl.delegate = Some(new TelemetrySpi {
+        override def client: BidiFlow[HttpRequest, HttpRequest, HttpResponse, HttpResponse, NotUsed] =
+          BidiFlow.identity
+
+        override def serverBinding: Flow[Tcp.IncomingConnection, Tcp.IncomingConnection, NotUsed] = Flow[Tcp.IncomingConnection]
+          .mapMaterializedValue { _ =>
+            probe.ref ! "bind-seen"
+            NotUsed
+          }
+          .map { conn =>
+            val connId = ConnectionId(UUID.randomUUID().toString)
+            probe.ref ! "connection-seen"
+            probe.ref ! connId
+            conn.copy(flow = conn.flow.addAttributes(Attributes(connId)))
+          }.watchTermination() { (notUsed, done) =>
+            done.onComplete(_ => probe.ref ! "unbind-seen")(system.dispatcher)
+            notUsed
+          }
+
+        override def serverConnection: BidiFlow[HttpResponse, HttpResponse, HttpRequest, HttpRequest, NotUsed] = BidiFlow.fromFlows(
+          Flow[HttpResponse].map { response =>
+            probe.ref ! "response-seen"
+            response
+          }.watchTermination() { (_, done) =>
+            done.foreach(_ => probe.ref ! "close-seen")(system.dispatcher)
+          },
+          StreamUtils.statefulAttrsMap[HttpRequest, HttpRequest](attributes =>
+            { request =>
+              probe.ref ! "request-seen"
+              attributes.get[ConnectionId].foreach(probe.ref ! _)
+              request
+            }
+          ))
+      })
+
+      val (serverBinding, http2ClientFlow) = bindAndConnect(probe)
+      probe.expectMsg("bind-seen")
+
+      val responseProbe = TestProbe()
+      val (requestQueue, _) =
+        Source.queue(10, OverflowStrategy.fail)
+          .viaMat(http2ClientFlow)(Keep.left)
+          .toMat(Sink.actorRef(responseProbe.ref, "done"))(Keep.both)
+          .run()
+      requestQueue.offer(HttpRequest())
+
+      probe.expectMsg("connection-seen")
+      val connId = probe.expectMsgType[ConnectionId]
+
+      probe.expectMsg("request-seen")
+      probe.expectMsgType[ConnectionId] should ===(connId)
```

Comment:
Ok, so this tests that the attribute is propagated from the connection stream to the request stream.